### ISSUE TITLE
segfault on connect

### ISSUE THREAD

**Issue by Ustilago** [theinvisible/openfortigui#27]
Hi,

I just installed openfortigui_0.4.2-1_amd64_debian9.deb on my Debian Sid, and I can't connect. The GUI doesn't show any error, nor does the log files. But in /var/log/messages I have :

openfortigui[17653]: segfault at 20 ip 000055c8b7d2ad1b sp 00007ffc67b49bd0 error 4 in openfortigui[55c8b7cf1000+c4000]

If you need more informations, just tell me how to obtain them. I tried using gdb, but it doesn't seem to catch the segfault.

Regards,

Sébastien

**Comment by theinvisible** [comment]
Hi,

OpenFortiGUI spawns new processes for every VPN connection profile.

You can also start this VPN connection on command line and so use gdb. For example: sudo /usr/bin/openfortigui --start-vpn --vpn-name <yourvpnname> --main-config '/home/<yourusername>/.openfortigui/main.conf'

Just replace <yourvpnname> with your vpn profile name and <yourusername> with your Linux username.

You have also checked the separate vpn connection logs? Just right-click your VPN profile and click "View logs".

Cheers Rene

**Comment by Ustilago** [comment]
Hi,

The log file is empty.

Here is gdb trace, but it lacks debug symbols :

gdb --args /usr/bin/openfortigui --start-vpn --vpn-name --main-config main.conf
[...]
Reading symbols from /usr/bin/openfortigui...(no debugging symbols found)...done.
(gdb) run
Starting program: /usr/bin/openfortigui --start-vpn --vpn-name --main-config main.conf
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x000055555558dd1b in ?? ()
(gdb) bt
#0  0x000055555558dd1b in ?? ()
#1  0x000055555558e633 in ?? ()
#2  0x0000555555567b54 in ?? ()
#3  0x00007ffff4974a87 in __libc_start_main (main=0x555555567390, argc=5, argv=0x7fffffffec08, init=<optimized out>, fini=<optimized out>, 
    rtld_fini=<optimized out>, stack_end=0x7fffffffebf8) at ../csu/libc-start.c:310
#4  0x00005555555685ba in ?? ()
(gdb) 

Regards,

Sébastien

**Comment by theinvisible** [comment]
Hi,

this doesnt really help.

Is it possible for you to build a debug build?

Just follow https://hadler.me/linux/openfortigui/ but replace "qmake" for "qmake openfortigui-project.pro -spec linux-g++ CONFIG+=debug CONFIG+=qml_debug && /usr/bin/make qmake_all"

Thanks and best regards
Rene

**Comment by Ustilago** [comment]
Hi,

Here is the result : 

Program received signal SIGSEGV, Segmentation fault.
0x000055555556b468 in QString::isEmpty (this=0x20) at ../../../../include/x86_64-linux-gnu/qt5/QtCore/qstring.h:926
926     { return d->size == 0; }
(gdb) bt
#0  0x000055555556b468 in QString::isEmpty (this=0x20) at ../../../../include/x86_64-linux-gnu/qt5/QtCore/qstring.h:926
#1  0x000055555559170e in vpnProcess::startVPN (this=0x7fffffffe4b0) at proc/vpnprocess.cpp:101
#2  0x00005555555912b5 in vpnProcess::run (this=0x7fffffffe4b0, vpnname=...) at proc/vpnprocess.cpp:45
#3  0x000055555556a239 in main (argc=5, argv=0x7fffffffebf8) at main.cpp:148

Regards,

Sébastien

**Comment by theinvisible** [comment]
Thanks for your trace.

Do you have username or password or both set in your vpn profile or all empty?

Regards
Rene

**Comment by Ustilago** [comment]
Both are in the vpn profile, I do the same on my laptop, and it works.

I misread your first message, so the backtrace was not correct : I launched openfortivpn as root, but with no vpn profile.

So I launched it as root to create a vpn profile, without password.

If I launch from command line, with the profile name, and the correct path for main.conf, it doesn't ask for a password, takes some time and ends with a segmentation fault :
Program received signal SIGSEGV, Segmentation fault.
0x000055555559ce3e in vpnWorker::end (this=0x7fffffffe628) at proc/vpnworker.cpp:467
467         ptr_tunnel->on_ppp_if_down(ptr_tunnel);
(gdb) bt
#0  0x000055555559ce3e in vpnWorker::end (this=0x7fffffffe628) at proc/vpnworker.cpp:467
#1  0x00005555555913e8 in vpnProcess::closeProcess (this=0x7fffffffe4c0) at proc/vpnprocess.cpp:52
#2  0x00005555555917bc in vpnProcess::startVPN (this=0x7fffffffe4c0) at proc/vpnprocess.cpp:117
#3  0x00005555555912b5 in vpnProcess::run (this=0x7fffffffe4c0, vpnname=...) at proc/vpnprocess.cpp:45
#4  0x000055555556a239 in main (argc=6, argv=0x7fffffffec08) at main.cpp:148

I also realised that I was using 0.4.0 version as my normal user, launching 0.6.0 version from git doesn't produce a segmentation fault, it just ends up with this in the log :
INFO:   Closed connection to gateway.
ERROR:  connect: Connection timed out
INFO:   Closed connection to gateway.
ERROR:  connect: Connection timed out
INFO:   Could not log out.

At the same time I can connect to the same fortinet from my laptop. Both computers are using Debian Sid, and are on the same network.

Regards,

Sébastien

**Comment by theinvisible** [comment]
Yes, its best to always test with the latest git version. Since 0.4.0 there were a lot of bugfixes and also version 0.6.0 is using a far newer openfortivpn core version. 

I must also say i didnt really focus on using the cmd version. If you want to test with the cmd method you must always set username and password in the VPN profile file because there is no function now implemented to query the user for the username/password on terminal.

Also the cmd only version still tries to connect to the (in this case non existing) GUI process so the errors with version 0.6.0 seems to come from the client trying to communicate via IPC. Maybe i will fix this in the future. But still the vpn connection should be usable.

If its working on another computer you can also try to check and compare the configuration on both computers. You can also check with "ldd" the used libraries on both computers. Another method is to try to connect with plain openfortivpn: https://github.com/adrienverge/openfortivpn 

Thanks and best regards
Rene

**Comment by theinvisible** [comment]
The crash on Ctrl+C should be fixed on 0.6.1 and latest git release.

As i said once i have not tested the detached/cmd only mode really. But thanks for your input, i will have a look at them.

**Comment by Ustilago** [comment]
Hi,

I've tried openfortivpn, it complains about a "Gateway certificate validation failed" error. With --trusted-cert XXX option, it connects correctly to the VPN.

In openfortigui I tried adding the sha256 digest in "Trusted-Cert (sha256)" in VPN-preferences, but it still fails to connect with :
INFO: Closed connection to gateway.
ERROR: connect: Connection timed out
INFO: Could not log out.

I forgot yesterday to check linked libraries on my home computer, I'll try to do it tonight or tomorrow, to see if there are differencies with my laptop.

Best Regards,

Sébastien
